core/src/main/scala/akka/persistence/cassandra/query/EventsByTagStage.scala

```diff
@@ -126,59 +149,91 @@ import com.datastax.driver.core.utils.UUIDs
         case _ =>
           throw new IllegalStateException("EventsByTagStage requires ActorMaterializer")
       }
 
+      private def calculateToOffset(): UUID = {
+        val to: Long = UUIDs.unixTimestamp(UUIDs.timeBased()) - settings.eventsByTagEventualConsistency.toMillis
+        val tOff = if (to < toOffsetMillis) {
+          // The eventual consistency delay is before the end of the query
+          val u = UUIDs.endOf(to)
+          if (log.isDebugEnabled) {
+            log.debug("{}: New toOffset (EC): {}", stageUuid, formatOffset(u))
+          }
+          u
+        } else {
+          // The eventual consistency delay is after the end of the query
+          if (log.isDebugEnabled) {
+            log.debug("{}: New toOffset (End): {}", stageUuid, formatOffset(toOffset.get))
+          }
+          toOffset.get
+        }
+        tOff
+      }
+
+      private def updateToOffset(): Unit = {
+        updateStageState(_.copy(toOffset = calculateToOffset()))
+      }
+
+      private def updateQueryState(state: QueryState): Unit = {
+        updateStageState(_.copy(state = state))
+      }
+
       implicit def ec: ExecutionContextExecutor = materializer.executionContext
+
       setHandler(out, this)
 
       val newResultSetCb = getAsyncCallback[Try[ResultSet]] {
         case Success(rs) =>
-          val _ = state match {
+          val _ = stageState.state match {
             case q: QueryInProgress => q
             case _ =>
-              throw new IllegalStateException(s"New ResultSet when in unexpected state $state")
+              throw new IllegalStateException(s"New ResultSet when in unexpected state ${stageState.state}")
           }
-          state = QueryResult(rs)
+          updateStageState(_.copy(state = QueryResult(rs)))
           tryPushOne()
         case Failure(e) =>
           log.warning("Cassandra query failed", e)
           fail(out, e)
       }
 
       override def preStart(): Unit = {
-        if (log.isDebugEnabled)
-          log.debug(
-            "Starting with initial pid tag sequence numbers: {}. Offset: {}. FromOffset: {}. To offset: {}",
-            tagPidSequenceNrs, usingOffset, formatOffset(fromOffset), toOffset.map(formatOffset)
-          )
+        stageState = StageState(
+          QueryIdle,
+          fromOffset,
+          calculateToOffset(),
+          initialTagPidSequenceNrs,
+          None,
+          bucketSize)
+        log.debug("[{}]: EventsByTag query starting with EC delay {}ms: {} ", stageUuid, settings.eventsByTagEventualConsistency.toMillis, stageState)
 
         if (settings.pubsubNotification) {
           Try {
             getStageActor {
               case (_, publishedTag) =>
                 if (publishedTag.equals(session.tag)) {
-                  log.debug("Received pub sub tag update for our tag, initiating query")
+                  log.debug("[{}] Received pub sub tag update for our tag, initiating query", stageUuid)
```

Comment:
how will the pub/sub notifications work with the EC delay? is the notification always arriving before the EC delay and will never retrieve the event until next poll?

I think the old (0.61) implementation had something to delay the notifications related to the EC delay.

Also, we have some similar notifications from Multi-DC Persistence.

Reply:
Yep looking at it the `EventByTagPublisher` scheduled the message to Continue after the EC. I'll do the same